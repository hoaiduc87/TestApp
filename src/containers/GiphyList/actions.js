import {
  GET_GIFS_TRENDING_REQUEST,
  GET_GIFS_TRENDING_SUCCESS,
  GET_GIFS_TRENDING_FAILURE,

  SET_GIFS_TRENDING_ANALYTICS,
} from './constants';
// GET GIFS TRENDING ACTIONS
export const getGifsTrendingRequest = (params) => ({
  type: GET_GIFS_TRENDING_REQUEST, params
});

export const getGifsTrendingSuccess = (data, isUpdate) => ({
  type: GET_GIFS_TRENDING_SUCCESS, data, isUpdate
});

export const getGifsTrendingFailure = (error) => ({
  type: GET_GIFS_TRENDING_FAILURE, error
});

export const setGifsTrendingAnalytics = (data, isUpdate) => ({
  type: SET_GIFS_TRENDING_ANALYTICS, data, isUpdate
});
