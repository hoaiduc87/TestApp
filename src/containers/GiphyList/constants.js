// API PATH
export const GET_GIFS_TRENDING_PATH = "gifs/trending";

// GET GIFS TRENDING CONSTANTS
export const GET_GIFS_TRENDING_REQUEST = "GET_GIFS_TRENDING_REQUEST";
export const GET_GIFS_TRENDING_SUCCESS = "GET_GIFS_TRENDING_SUCCESS";
export const GET_GIFS_TRENDING_FAILURE = "GET_GIFS_TRENDING_FAILURE";

export const SET_GIFS_TRENDING_ANALYTICS = "SET_GIFS_TRENDING_ANALYTICS";
