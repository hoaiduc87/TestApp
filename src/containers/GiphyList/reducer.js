import produce from 'immer';

import {
  GET_GIFS_TRENDING_REQUEST,
  GET_GIFS_TRENDING_SUCCESS,
  GET_GIFS_TRENDING_FAILURE,

  SET_GIFS_TRENDING_ANALYTICS,
} from './constants';

export const initialState = {
  trendingGif: {
    analytics: {},
    data: {},
    initloading: false,
    updateLoading: false,
    error: null,
  }
};

const reducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case GET_GIFS_TRENDING_REQUEST:
        if (action.params.isUpdate) {
          draft.trendingGif.updateLoading = true;
          draft.trendingGif.error = null;
        } else {
          draft.trendingGif.data = {};
          draft.trendingGif.initloading = true;
          draft.trendingGif.error = null;
        }
        break;

      case GET_GIFS_TRENDING_SUCCESS:
        if (action.isUpdate) {
          draft.trendingGif.data.data = [
            ...(draft.trendingGif.data.data || []),
            ...(action.data.data || [])
          ];
          draft.trendingGif.data.pagination = action.data.pagination;
          draft.trendingGif.updateLoading = false;
        } else {
          draft.trendingGif.data = action.data;
          draft.trendingGif.initloading = false;
        }
        break;

      case GET_GIFS_TRENDING_FAILURE:
        draft.trendingGif.error = action.error;
        draft.trendingGif.initloading = false;
        draft.trendingGif.updateLoading = false;
        break;

      case SET_GIFS_TRENDING_ANALYTICS:
        if (action.isUpdate) {
          draft.trendingGif.analytics = {
            ...draft.trendingGif.analytics,
            ...action.data
          };
        } else {
          draft.trendingGif.analytics = action.data;
        }
        break;

      default:
        break;
    }
  });

export default reducer;
