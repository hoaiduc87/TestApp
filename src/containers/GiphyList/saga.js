import { takeLatest, put, call, all } from 'redux-saga/effects';
import { apiGet } from 'services';
import { GIPHY_KEY } from 'settings';
import { requestMiddleware } from 'services/middlewares';

import {
  GET_GIFS_TRENDING_PATH,
  GET_GIFS_TRENDING_REQUEST,
} from './constants';

import {
  getGifsTrendingSuccess,
  getGifsTrendingFailure,
  setGifsTrendingAnalytics,
} from './actions';

function* getGifsTrending({ params }) {
  const { isUpdate, ...remainParams } = params;
  try {
    const { data } = yield requestMiddleware(call(apiGet, {
      path: GET_GIFS_TRENDING_PATH,
      body: {
        api_key: GIPHY_KEY,
        ...remainParams,
      }
    }));

    const currentTimestamp = Date.now();
    const loadAnalyticsRequest = {};
    const clickAnalyticsRequest = {};
    const sentAnalyticsRequest = {};
    const analyticsResults = {};

    if (data && data.data) {
      data.data.forEach(gif => {
        const analytics = gif.analytics || {};
        if ((analytics.onload || {}).url) {
          loadAnalyticsRequest[gif.id] = requestMiddleware(call(apiGet, {
            url: analytics.onload.url,
            body: {
              api_key: GIPHY_KEY,
              random_id: params.random_id,
              ts: currentTimestamp,
            }
          }));
        }
        if ((analytics.onclick || {}).url) {
          clickAnalyticsRequest[gif.id] = requestMiddleware(call(apiGet, {
            url: analytics.onclick.url,
            body: {
              api_key: GIPHY_KEY,
              random_id: params.random_id,
              ts: currentTimestamp,
            }
          }));
        }
        if ((analytics.onsent || {}).url) {
          sentAnalyticsRequest[gif.id] = requestMiddleware(call(apiGet, {
            url: analytics.onsent.url,
            body: {
              api_key: GIPHY_KEY,
              random_id: params.random_id,
              ts: currentTimestamp,
            }
          }));
        }
      });

      const analyticsReponses = yield all([
        all(loadAnalyticsRequest),
        all(clickAnalyticsRequest),
        all(sentAnalyticsRequest),
      ]);

      data.data.forEach(gif => {
        analyticsResults[gif.id] = {
          load: analyticsReponses[0][gif.id].data || 0,
          click: analyticsReponses[1][gif.id].data || 0,
          sent: analyticsReponses[2][gif.id].data || 0,
        };
      });
    }

    yield put(setGifsTrendingAnalytics(analyticsResults, isUpdate));
    yield put(getGifsTrendingSuccess(data || {}, isUpdate));
  } catch (ex) {
    yield put(setGifsTrendingAnalytics({}));
    yield put(getGifsTrendingFailure(ex));
  }
}

export default function* giphyListSaga() {
  yield takeLatest(GET_GIFS_TRENDING_REQUEST, getGifsTrending);
}
