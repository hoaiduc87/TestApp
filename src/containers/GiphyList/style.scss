@import 'assets/styles/mixins.scss';

.giphy-trending {
  &__container {
    position: relative;
    padding: 16px;
    min-height: inherit;
  }

  &__list {
    display: grid;
    grid-row-gap: 24px;
    grid-template-columns: 23% 23% 23% 23%;
    justify-content: space-around;

    @include tablet {
      grid-row-gap: 16px;
      grid-template-columns: 31% 31% 31%;
    }

    @include mobile {
      grid-row-gap: 8px;
      grid-template-columns: 46% 46%;
    }
  }

  &__modal__body {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }

  &__footer {
    position: relative;
    height: 100px;
    text-align: center;
  }
}
