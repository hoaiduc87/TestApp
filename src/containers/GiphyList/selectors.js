import { createSelector } from 'reselect';
import { initialState } from './reducer';

const giphyListState = state => state.giphyList || initialState;

const selectGifsTrendingData = () =>
  createSelector(
    giphyListState,
    substate => ({
      ...substate.trendingGif.data,
      data: (substate.trendingGif.data.data || []).map(gif => ({
        ...gif,
        analytics: substate.trendingGif.analytics[gif.id] || {}
      }))
    })
  );

const selectGifsTrendingInitLoading = () =>
  createSelector(
    giphyListState,
    substate => substate.trendingGif.initloading
  );

const selectGifsTrendingUpdateLoading = () =>
  createSelector(
    giphyListState,
    substate => substate.trendingGif.updateLoading
  );

const selectGifsTrendingError = () =>
  createSelector(
    giphyListState,
    substate => substate.trendingGif.error
  );

export {
  selectGifsTrendingData,
  selectGifsTrendingInitLoading,
  selectGifsTrendingUpdateLoading,
  selectGifsTrendingError,
};
