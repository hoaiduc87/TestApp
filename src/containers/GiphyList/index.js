/* eslint-disable react-hooks/exhaustive-deps */
import React, { memo, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import GiphyCard from 'components/GiphyCard';
import Modal from 'components/Modal';
import Loading from 'components/Loading';

import { selectRandomId } from 'containers/App/selectors';

import { getGifsTrendingRequest } from './actions';
import {
  selectGifsTrendingData,
  selectGifsTrendingInitLoading,
  selectGifsTrendingUpdateLoading,
  selectGifsTrendingError,
} from './selectors';

import './style.scss';

const ITEMS_PER_PAGE = 20;

function GiphyList({
  randomId,
  gifsTrending,
  initLoading,
  updateLoading,
  error,
  getGifsTrending,
}) {
  let resizeTimeout = null;
  let scrollTimeout = null;
  let isLoadingMore = false;
  const gifs = gifsTrending.data || [];
  const pagination = gifsTrending.pagination || {};
  const canLoadMore = pagination.total_count > pagination.offset + pagination.count;

  const [isScrollable, setIsScrollable] = useState(false);
  const [modal, setModal] = useState({});

  useEffect(() => {
    getGifsTrending({
      random_id: randomId,
      rating: 'g',
      limit: ITEMS_PER_PAGE,
    });
  }, []);

  useEffect(() => {
    handleResize();
  }, [gifs.length]);

  useEffect(() => {
    document.addEventListener('scroll', handleScroll);
    return () => {
      document.removeEventListener('scroll', handleScroll);
    }
  }, [isScrollable, gifs.length, canLoadMore, error]);

  useEffect(() => {
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    }
  }, [isScrollable]);

  const handleScroll = () => {
    if (!isScrollable || !canLoadMore || isLoadingMore || error) {
      return;
    }

    if (scrollTimeout) {
      clearTimeout(scrollTimeout);
    }

    scrollTimeout = setTimeout(() => {
      const scrollRemain = document.documentElement.scrollHeight - document.documentElement.scrollTop - document.documentElement.clientHeight;

      if (scrollRemain < document.documentElement.clientHeight) {
        isLoadingMore = true;
        getGifsTrending({
          random_id: randomId,
          rating: 'g',
          limit: ITEMS_PER_PAGE,
          offset: pagination.offset + pagination.count,
          isUpdate: true,
        });
      }
    }, 100);
  };

  const handleLoadMore = () => {
    getGifsTrending({
      random_id: randomId,
      rating: 'g',
      limit: ITEMS_PER_PAGE,
      offset: pagination.offset + pagination.count,
      isUpdate: true,
    });
  };

  const handleResize = () => {
    if (resizeTimeout) {
      clearTimeout(resizeTimeout);
    }

    resizeTimeout = setTimeout(() => {
      const tmpOffset = document.documentElement.offsetHeight - document.documentElement.clientHeight;

      if (!isScrollable && document.documentElement.clientHeight < tmpOffset) {
        setIsScrollable(true);
      }

      if (isScrollable && document.documentElement.clientHeight >= tmpOffset) {
        setIsScrollable(false);
      }
    }, 100);
  };

  return (
    <div className="giphy-trending__container">
      <Modal
        show={!!modal.imageSrc}
        className="modal--fullscreen"
        onClose={() => setModal({})}
      >
        <img className="giphy-trending__modal__body" src={modal.imageSrc} alt="fullscreen" />
      </Modal>
      <Loading loading={initLoading} />
      <div className="giphy-trending__list">
        {gifs.map((gifItem, idx) =>
          <GiphyCard
            key={`${gifItem.id}-${idx}`}
            data={gifItem || {}}
            onClickImage={(imageSrc) => setModal({ imageSrc })}
          />
        )}
      </div>
      <div className="giphy-trending__footer">
        <Loading loading={updateLoading} className="loading--transparent" width="60px" />
        {(!isScrollable || error) && canLoadMore && !updateLoading &&
          <button
            type="button"
            className="default-btn default-btn--primary default-btn--round"
            onClick={handleLoadMore}
          >See more gifs</button>
        }
      </div>
    </div>
  )
}

GiphyList.propTypes = {
  randomId: PropTypes.string,
  gifsTrending: PropTypes.object,
  initLoading: PropTypes.bool,
  updateLoading: PropTypes.bool,
  error: PropTypes.any,
  getGifsTrending: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  randomId: selectRandomId(),
  gifsTrending: selectGifsTrendingData(),
  initLoading: selectGifsTrendingInitLoading(),
  updateLoading: selectGifsTrendingUpdateLoading(),
  error: selectGifsTrendingError(),
});

const mapDispatchToProps = (dispatch) => ({
  getGifsTrending: params => dispatch(getGifsTrendingRequest(params)),
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  withConnect,
  memo,
)(GiphyList);
