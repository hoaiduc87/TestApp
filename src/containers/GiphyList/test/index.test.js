import React from 'react';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import GiphyList from '..';

const mockStore = configureMockStore();
const initialState = {
  giphyList: {
    trendingGif: {
      analytics: {},
      data: {
        data: [
          {

          }
        ],
        pagination: {
          total_count: 100,
          offset: 0,
          count: 1
        }
      },
      initloading: false,
      updateLoading: false,
      error: null,
    }
  }
};
const store = mockStore(initialState);

describe('<GiphyList/>', () => {
  it('Should run and match snapshot', () => {
    const Wrapper = ({children}) => <Provider store={store}>{children}</Provider>;
    const {
      container: { firstChild }
    } = render(<GiphyList />, { wrapper: Wrapper });
    expect(firstChild).toMatchSnapshot();
  });

  it('Should execute setNotifyModalRequest once click close button', () => {
    store.dispatch = jest.fn();
    const comp = mount(
      <Provider store={store}>
        <GiphyList />
      </Provider>
    );
    comp.find('.default-btn--primary').first().simulate('click');
    expect(store.dispatch).toHaveBeenCalledTimes(2);
  });
})