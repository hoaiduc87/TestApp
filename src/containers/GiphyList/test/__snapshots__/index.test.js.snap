// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiphyList/> Should run and match snapshot 1`] = `
<div
  class="giphy-trending__container"
>
  <div
    class="modal modal--hidden"
  >
    <div
      class="modal__dialog modal--fullscreen"
    >
      <span
        class="modal__dialog__close"
      >
        <img
          alt="close"
          src="close_white.svg"
        />
      </span>
      <img
        alt="fullscreen"
        class="giphy-trending__modal__body"
      />
    </div>
    <div
      class="modal__backdrop"
    />
  </div>
  <div
    class="loading  hidden"
  >
    <img
      alt="loading"
      src="loading.gif"
      width="100px"
    />
  </div>
  <div
    class="giphy-trending__list"
  >
    <div
      class="giphy-item"
    >
      <div
        class="giphy-card"
      >
        <div
          class="giphy-card__image"
        >
          <img
            alt="thumb"
          />
        </div>
        <div
          class="giphy-card__footer"
        >
          <div
            class="giphy-card__footer--left"
          >
            <img
              alt=""
              class="analytics-attachment-icon"
              src="attachment.svg"
            />
          </div>
          <div
            class="giphy-card__footer--right"
          >
            <span
              class="analytics-item"
            >
              <img
                alt=""
                src="eye.svg"
              />
              <span>
                0
              </span>
            </span>
            <span
              class="analytics-item"
            >
              <img
                alt=""
                src="chat_bubble.svg"
              />
              <span>
                0
              </span>
            </span>
            <span
              class="analytics-item"
            >
              <img
                alt=""
                src="heart.svg"
              />
              <span>
                0
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="giphy-trending__footer"
  >
    <div
      class="loading loading--transparent hidden"
    >
      <img
        alt="loading"
        src="loading.gif"
        width="60px"
      />
    </div>
    <button
      class="default-btn default-btn--primary default-btn--round"
      type="button"
    >
      See more gifs
    </button>
  </div>
</div>
`;
